### PR TITLE
Classification logistic regression

### PR DESCRIPTION
**Pull request by HannahPetry**
# PR Description

## Please check if the PR fulfills these requirements

- [ ] Tests for the changes have been added (for bug fixes/features) (Or not needed)
- [x] Labels, reviewers and assignees have been added to the pull request
- [ ] Docs have been added / updated (for bug fixes / features) (Or not needed)

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -127,3 +127,6 @@ dmypy.json
 
 # Pyre type checker
 .pyre/
+
+# data
+data/en_ner_bionlp13cg_md-0.5.1/
```
Comment (florentinvonhaugwitz) [inline .gitignore:139]: Können wir diesen Ordner in einen subfolder von data verschieben? Am besten data/external/en_ner_biolon13cg_md-0.5.1/

Comment (florentinvonhaugwitz) [inline .gitignore:139]: Ich sehe außerdem die ganzen files aus diesem Ordner auch im PR. Kannst du die wieder rausnehmen damit wir die nicht hochladen am Ende?

File: src/models/predict_lr_model.py [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+def load_model() -> object:
+    """
+    Load model from disk
+
+    Returns
+    -------
+    object
+        model
+    """
+    model_name = "./models/sklearn_logistic_regression_model.pkl"
```
Comment (florentinvonhaugwitz) [inline src/models/predict_lr_model.py]: Hier würde ich mit os.path.join() arbeiten, damit wir sicherstellen können, dass es auf allen Computern funktioniert

File: src/models/predict_lr_model.py [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+    model = joblib.load(filename=model_name)
```
Comment (florentinvonhaugwitz) [inline src/models/predict_lr_model.py]: Ist joblib besser als `pickle.dump(model, open('model.pkl', 'wb'))`?

File: src/models/predict_lr_model.py (hunk 2 above)
Comment (florentinvonhaugwitz) [inline src/models/predict_lr_model.py]: Vielleicht können wir auch model_name als Paramter für die Funktion nutzen und dann können wir diese Funktion generisch für andere Modell auch benutzen

File: src/models/predict_lr_model.py [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+
+
+def predict_probability(model: object, value: str, category_list: list) -> pd.DataFrame:
+    """
+    get probabilities for sample
+
+    Parameters
+    ----------
+    model : object
+        model
```
Comment (florentinvonhaugwitz) [inline src/models/predict_lr_model.py]: Hier am besten noch beschreiben was das genau für ein Parameter sit

File: src/models/predict_lr_model.py [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+# load_model()
+# predict_value()
+
+
+import joblib
+import pandas as pd
+import numpy as np
+
+
+def load_model() -> object:
```
Comment (florentinvonhaugwitz) [inline src/models/predict_lr_model.py]: In dem Fall wird nicht ein object sondern ein imblearn.pipeline.Pipeline Wert zurückgegeben

File: src/models/predict_lr_model.py [hunk 6]
```diff
@@ -0,0 +1,74 @@
...
+    -------
+    object
+        model
+    """
+    model_name = "./models/sklearn_logistic_regression_model.pkl"
+    model = joblib.load(filename=model_name)
+    return model
+
+
+def predict_probability(model: object, value: str, category_list: list) -> pd.DataFrame:
```
Comment (florentinvonhaugwitz) [inline src/models/predict_lr_model.py]: Hier muss auch object auch wieder zu imblearn.pipeline.Pipeline
Und am besten auch angeben was für werte in list übergeben werden, also List[str] (https://stackoverflow.com/questions/24853923/type-hinting-a-collection-of-a-specified-type)

File: src/models/predict_lr_model.py [hunk 7]
```diff
@@ -0,0 +1,74 @@
...
+    """
+    get probabilities for sample
+
+    Parameters
+    ----------
+    model : object
```
Comment (florentinvonhaugwitz) [inline src/models/predict_lr_model.py]: Hier wieder imblearn.pipeline.Pipeline

File: src/models/predict_lr_model.py [hunk 8]
```diff
@@ -0,0 +1,74 @@
...
+    prob_df = pd.DataFrame(
+        prob_array, index=category_list, columns=["Probability"]
+    ).sort_values(by="Probability", ascending=False)
+    return prob_df
+
+
+def main():
+    # Load model
+    model = load_model()
+    print(model)
```
Comment (florentinvonhaugwitz) [inline src/models/predict_lr_model.py]: Das können wir rausnehmen

File: src/models/predict_lr_model.py [hunk 9]
```diff
@@ -0,0 +1,74 @@
...
+        " SOAP / Chart / Progress Notes"
+        " Radiology"
+        " Orthopedic"
+        " Obstetrics / Gynecology"
+        " Neurology"
+        " Gastroenterology"
+        " Consult - History and Phy."
+    ]
+    prob_df = predict_probability(model, value, category_list)
+    print(prob_df)
```
Comment (florentinvonhaugwitz) [inline src/models/predict_lr_model.py]: Ist das einfach ein Beispielwert zur Veranschaulichung?

File: src/models/train_lr_model.py [hunk 10]
```diff
@@ -0,0 +1,256 @@
...
+    -------
+    pd.DataFrame
+        dataframe with labels and NLP features
+    """
+    df = pd.read_csv(file_path)
+    return df
+
+
+# Retrieve labels as function
+def get_labels(df: pd.DataFrame) -> list:
```
Comment (florentinvonhaugwitz) [inline src/models/train_lr_model.py]: Hier auch angeben was für einen art von Liste es ist

File: src/models/train_lr_model.py [hunk 11]
```diff
@@ -0,0 +1,256 @@
...
+    Returns
+    -------
+    list
+        list of all labels
+    """
+    return df["medical_specialty"].tolist()
+
+
+# Split the dataframe into test and train data
+def split_data(df: pd.DataFrame) -> array:
```
Comment (florentinvonhaugwitz) [inline src/models/train_lr_model.py]: Der Rückgabewert ist glaube ich ein Tupel von np.arrays

File: src/models/train_lr_model.py [hunk 12]
```diff
@@ -0,0 +1,256 @@
...
+    y = get_labels(df)
+    X_train, X_test, y_train, y_test = train_test_split(
+        X, y, test_size=0.2, random_state=42
+    )
+
+    return X_train, X_test, y_train, y_test
+
+
+# Build pipeline
+def build_pipeline() -> imbPipeline:
```
Comment (florentinvonhaugwitz) [inline src/models/train_lr_model.py]: Hier ist der Rückgabewert glaube ich auch imblearn.pipeline.Pipeline

File: src/models/train_lr_model.py [hunk 13]
```diff
@@ -0,0 +1,256 @@
...
+        fitted model
+    """
+    model.fit(X_train, y_train)
+    return model
+
+
+# Grid search
+def grid_search(
+    X_train: array, y_train: array, model_pipeline: imbPipeline, param_grid: list
+) -> GridSearchCV:
```
Comment (florentinvonhaugwitz) [inline src/models/train_lr_model.py]: Müsste der Rückgabewert nicht auch ein Model sein, da wir ja den besten estimator zurückgeben?
Am besten einmal `print(type(search.best_estimator_))` ausführen vor dem return wert um zu sehen was er da ausgibt

**Review by florentinvonhaugwitz — COMMENTED** [review]
Das ist richtiger guter Hannah!! Sehr organisiert und strukturiert. Ich habe ein paar Anmerkungen bei kleineren Sachen aber die Logik erscheint mir richtig und umfänglich! Falls du ein paar meiner Kommentare anders siehst, dann schreib es einfach dazu - letzlich sind es ja nur Anregungen

File: src/models/train_lr_model.py (hunk 10 above)
Comment (HannahPetry) [reply src/models/train_lr_model.py:48]: Mein type(get_labels(df_test)) ist einfach list. Kann man das noch spezifizieren oder meinst du im #Titel der Funktion?

File: src/models/predict_lr_model.py (hunk 9 above)
Comment (HannahPetry) [reply src/models/predict_lr_model.py]: Ja, genau. Das funktioniert momentan nur leider noch nicht mit einem value.

**Review by florentinvonhaugwitz — COMMENTED** [review]
Am besten alle "test" models im Ordner models/clf löschen

File: src/clf/create_data.py [hunk 14]
```diff
@@ -0,0 +1,66 @@
...
+    df[column_name] = df[column_name].apply(lambda x: replace_tab(x))
+    df[column_name] = df[column_name].apply(lambda x: " ".join(x))
+    return df
+
+
+def main():
+    # Load data
+    file_path = os.path.join(
+        "data", "processed", "nlp", "mtsamples", "mtsamples_unsupervised_both_v2.csv"
+    )
```
Comment (florentinvonhaugwitz) [inline src/clf/create_data.py]: Am besten ist es wenn du für alle Files oder sonstigen Konstanten einen "constants.py" anlegst der die alle beinhaltet

File: src/clf/create_data.py [hunk 15]
```diff
@@ -0,0 +1,66 @@
...
+        df.medical_specialty,
+        test_size=0.2,
+        random_state=42,
+    )
+
+    # Save data as csv
+    train_df = pd.DataFrame({"keywords": X_train, "medical_specialty": y_train})
+    print(train_df.head())
+    test_df = pd.DataFrame({"keywords": X_test, "medical_specialty": y_test})
+    print(test_df.head())
```
Comment (florentinvonhaugwitz) [inline src/clf/create_data.py]: Ist das zur Überprüfung?

File: src/clf/evaluation.py [hunk 16]
```diff
@@ -0,0 +1,186 @@
...
+import os
+import pickle
+from sklearn.metrics import confusion_matrix
+from sklearn.metrics import classification_report
+from sklearn.metrics import make_scorer
+from utils import load_data
+
+
+# Other metrics for model evaluation (accuracy @k optimized for and MRR @k)
+def _reciprocal_rank(true_labels: list, machine_preds: list) -> float:
```
Comment (florentinvonhaugwitz) [inline src/clf/evaluation.py]: Wieso hat die Funktion einen Unterstrich?

File: src/clf/evaluation.py [hunk 17]
```diff
@@ -0,0 +1,186 @@
...
+        # for RR we need position of first correct item
+        first_pos_list = tp_pos_list[0]
+
+        # rr = 1/rank
+        rr = 1 / float(first_pos_list)
+
+    return rr
+
+
+def compute_mrr_at_k(items: list) -> float:
```
Comment (florentinvonhaugwitz) [inline src/clf/evaluation.py]: Wenn der type list ist, dann am besten angeben was dort entalten ist. Also in diesem Fall list[tuples[str, str]]

File: src/clf/evaluation.py (hunk 16 above)
Comment (florentinvonhaugwitz) [inline src/clf/evaluation.py]: Die werte von list auch noch angeben

File: src/clf/evaluation.py [hunk 18]
```diff
@@ -0,0 +1,186 @@
...
+        list of tuples (true labels, machine predictions)
+    """
+
+    pred_gold_list = [[[Y_test.iloc[idx]], pred] for idx, pred in enumerate(Y_preds)]
+    return pred_gold_list
+
+
+def get_top_k_predictions(
+    model: imblearn.pipeline.Pipeline, X_test: pd.core.series.Series, k: int
+) -> list:
```
Comment (florentinvonhaugwitz) [inline src/clf/evaluation.py]: Wert für list angeben

File: src/clf/evaluation.py [hunk 19]
```diff
@@ -0,0 +1,186 @@
...
+    # evaluate model
+    for model in [lr_model]:
+        preds = get_top_k_predictions(model, X_test, 3)
+        eval_items = collect_preds(y_test, preds)
+
+        accuracy = compute_accuracy(eval_items)
+        mrr = compute_mrr_at_k(eval_items)
+
+        print("Accuracy @3: ", accuracy)
+        print("MRR @3: ", mrr)
```
Comment (florentinvonhaugwitz) [inline src/clf/evaluation.py]: Nice! Sehr gut!

File: src/clf/lr_training.py [hunk 20]
```diff
@@ -0,0 +1,182 @@
...
+import imblearn
+import os
+
+from sklearn.linear_model import LogisticRegression
+from sklearn.model_selection import GridSearchCV
+
+from sklearn.pipeline import Pipeline
+from imblearn.pipeline import Pipeline as imbPipeline
+
+from traitlets import List
```
Comment (florentinvonhaugwitz) [inline src/clf/lr_training.py:18]: Bist du sicher dass wir diesen import brauchen?

File: src/clf/lr_training.py [hunk 21]
```diff
@@ -0,0 +1,182 @@
...
+
+from utils import load_data
+from utils import preprocessing_pipeline
+
+# Build pipeline
+def build_pipeline(pipeline) -> imblearn.pipeline.Pipeline:
```
Comment (florentinvonhaugwitz) [inline src/clf/lr_training.py]: type von pipeline angeben

File: src/clf/lr_training.py [hunk 22]
```diff
@@ -0,0 +1,182 @@
...
+    """
+    Build pipeline for model
+
+    Parameters
+    ----------
+    preprocessing_pipeline : imblearn.pipeline.Pipeline
+        pipeline with preprocessing steps
```
Comment (florentinvonhaugwitz) [inline src/clf/lr_training.py:28]: Absatz

File: src/clf/lr_training.py [hunk 23]
```diff
@@ -0,0 +1,182 @@
...
+    )
+    search.fit(X_train, y_train)
+    print("Best parameters:", search.best_params_)
+    print("Best cross-validation score: {:.2f}".format(search.best_score_))
+    return search.best_estimator_
+
+
+def main():
+    # Load train data
+    train_file_path = os.path.join("data", "processed", "clf", "train.csv")
```
Comment (florentinvonhaugwitz) [inline src/clf/lr_training.py:143]: Den file path auch in den constants file schieben

File: src/clf/lr_training.py [hunk 24]
```diff
@@ -0,0 +1,182 @@
...
+    X_train, y_train = load_data(train_file_path)
+    print(type(X_train))
+    # Load test data
+    test_file_path = os.path.join("data", "processed", "clf", "test.csv")
```
Comment (florentinvonhaugwitz) [inline src/clf/lr_training.py:147]: Den file path auch in den constants file schieben

File: src/clf/lr_training.py [hunk 25]
```diff
@@ -0,0 +1,182 @@
...
+    y_train : pd.core.series.Series
+        train labels
+
+    Returns
+    -------
+    imblearn.pipeline.Pipeline
+        fitted model
+    """
+    model.fit(X_train, y_train)
+    return model
```
Comment (florentinvonhaugwitz) [inline src/clf/lr_training.py]: Im Zweifel kannst du diese Methode auch rausnehmen da sie nur eine Zeile lang ist. Da ist es schneller sie direkt in der main methode auszuführen

File: src/clf/predict.py [hunk 26]
```diff
@@ -0,0 +1,121 @@
...
+    -------
+    pd.Series
+        Top symptoms for each class
+    """
+
+    coef = model.named_steps["clf"].coef_
+    vectorizer = model.named_steps["vect"]
+    feat = vectorizer.get_feature_names()
+    coef_df = pd.DataFrame(coef, columns=feat, index=model.classes_)
+    coef_df = coef_df.abs()
```
Comment (florentinvonhaugwitz) [inline src/clf/predict.py]: Wieso machen wir nochmal .abs()?

File: src/streamlit/predict_lr_new.py [hunk 27]
```diff
@@ -0,0 +1,153 @@
...
+"""
+Description:
+    Predicting probability for medical labels from saved model 
+    Showing top features model uses per class in general
+    Showing features from sample the model uses to predict top classes 
+    
+Usage:
+    
+Possible arguments:
+    *
```
Comment (florentinvonhaugwitz) [inline src/streamlit/predict_lr_new.py]: Kann weg

File: src/streamlit/predict_lr_new.py [hunk 28]
```diff
@@ -0,0 +1,153 @@
...
+    model_name : str
+        path to saved model
+
+    Returns
+    -------
+    imblearn.pipeline.Pipeline
+        best model from train.py
+    """
+    model = pickle.load(open(file_path, "rb"))
+    return model
```
Comment (florentinvonhaugwitz) [inline src/streamlit/predict_lr_new.py]: Kannst du direkt implementieren ohne Funktion, da es nur eine Zeile ist

**Review by florentinvonhaugwitz — CHANGES_REQUESTED** [review]
Weiterhin sehr guter code! 
Meiner Meinung nach fehlen nur noch die paar Kleinigkeiten, die ich angemerkt habe, dann
am besten die main branch in deine branch ziehen (können wir gerne auch morgen zusammen machen) und `pre-commit run --all-files` einmal laufen lassen und gucken ob es durch geht. 
Die docs können wir gerne auch morgen zusammen machen

File: src/clf/predict.py (hunk 26 above)
Comment (HannahPetry) [reply src/clf/predict.py]: Das muss ich abändern, das war noch mit allen coefficients, auch negativen. Gut, dass du das gesehen hast. Man muss sich ja nur die größten positiven coefficients angeben lassen, richtig?

File: src/clf/create_data.py (hunk 15 above)
Comment (HannahPetry) [reply src/clf/create_data.py]: Ja, nehm ich raus.

**Review by florentinvonhaugwitz — DISMISSED** [review]

**Review by florentinvonhaugwitz — APPROVED** [review]
